### PR TITLE
Some refactor of expr sort and fix a bug for a matmul schedule

### PR DESCRIPTION
**Pull request by zasdfgbnm**
# The problem

I see an error in the following matmul schedule:
https://github.com/csarofeen/pytorch/compare/tracking-matmul...csarofeen:pytorch:matmul-schedule-test?expand=1
And I am attempting to fix it.

The problem I believe is: I think we should have another fallback mode where the following constraint is lifted.
```C++
  if (consumer_group->payload()->pa_domains.size() <
      consumer_group->payload()->ca_domains.size()) {
    return false;
  }
```

In a matmul schedule, we have
Screenshot_20230120_125544

And if at some point, we have
```
group1:  acw_smem = cp.async(a)
```
```
group 2: everything else except cc->c_smem->c
```
Then group 2's `pa_domain` will not have any ID about `N`, but group 2's `ca_domains` do have such IDs. In such case `pa_domains.size() < ca_domains.size()` but still, we need to merge group 1 and 2.

Related debugging print:
```
supportedMerge:
producer_group:Group Start{
  ca, pa (3, 0) ca_ids {iblockIdx.y95{( ceilDiv(( ceilDiv(T1.size[1], 128) ), 2) )}, iS96{2}, rS27{( ceilDiv(T0.size[0], 32) )}} pa_ids {}
Exprs {
T7_s[ iblockIdx.y79{( ceilDiv(( ceilDiv(T1.size[1], 128) ), 2) )}, iS80{2}, iS81{( ceilDiv(T0.size[0], 32) )}, iS177{( ceilDiv(( ( ( ( ceilDiv(32, 8) ) * ( ceilDiv(8, 1) ) ) * 1 ) * ( ( ( ceilDiv(128, 64) ) * ( ceilDiv(64, 8) ) ) * 8 ) ), 2048) )}, ithreadIdx.z183{( ceilDiv(( ceilDiv(( ceilDiv(2048, 8) ), 32) ), 2) )}, ithreadIdx.y184{2}, ithreadIdx.x182{32}, iV180{8} ] ca_pos( 3 ) = CpAsyncCg( T1_g[ iS85{( ceilDiv(( ceilDiv(T1.size[1], 128) ), 2) )}, iS86{2}, iS87{( ceilDiv(T0.size[0], 32) )}, iS84{128}, iS88{32} ] )
}Group End

consumer_group:Group Start{
  ca, pa (4, 3) ca_ids {iblockIdx.x91{( ceilDiv(( ceilDiv(T0.size[1], 256) ), 2) )}, iblockIdx.y95{( ceilDiv(( ceilDiv(T1.size[1], 128) ), 2) )}, iS92{2}, iS96{2}} pa_ids {iblockIdx.y95{( ceilDiv(( ceilDiv(T1.size[1], 128) ), 2) )}, iS96{2}, rS27{( ceilDiv(T0.size[0], 32) )}}
Exprs {
T6_s[ iblockIdx.x61{( ceilDiv(( ceilDiv(T0.size[1], 256) ), 2) )}, iS62{2}, iS63{( ceilDiv(T0.size[0], 32) )}, iS154{( ceilDiv(( ( ( ( ceilDiv(32, 8) ) * ( ceilDiv(8, 1) ) ) * 1 ) * ( ( ( ceilDiv(256, 64) ) * ( ceilDiv(64, 8) ) ) * 8 ) ), 2048) )}, ithreadIdx.z160{( ceilDiv(( ceilDiv(( ceilDiv(2048, 8) ), 32) ), 2) )}, ithreadIdx.y161{2}, ithreadIdx.x159{32}, iV157{8} ] ca_pos( 3 ) = CpAsyncCg( T0_g[ iS67{( ceilDiv(( ceilDiv(T0.size[1], 256) ), 2) )}, iS68{2}, iS69{( ceilDiv(T0.size[0], 32) )}, iS66{256}, iS70{32} ] )
T8_l[ iblockIdx.x55{( ceilDiv(( ceilDiv(T0.size[1], 256) ), 2) )}, iS56{2}, iS57{( ceilDiv(T0.size[0], 32) )}, iS127{( ceilDiv(32, 16) )}, ithreadIdx.z129{( ceilDiv(256, 64) )}, iS131{( ceilDiv(64, 16) )}, ithreadIdx.x202{( ( ( ceilDiv(16, 8) ) * ( ceilDiv(16, 8) ) ) * 8 )}, iV200{8} ] ca_pos( 4 ) produce_pos( 3 ) = LdMatrixTranspose( T6_s[ iblockIdx.x61{( ceilDiv(( ceilDiv(T0.size[1], 256) ), 2) )}, iS62{2}, iS63{( ceilDiv(T0.size[0], 32) )}, iS154{( ceilDiv(( ( ( ( ceilDiv(32, 8) ) * ( ceilDiv(8, 1) ) ) * 1 ) * ( ( ( ceilDiv(256, 64) ) * ( ceilDiv(64, 8) ) ) * 8 ) ), 2048) )}, ithreadIdx.z160{( ceilDiv(( ceilDiv(( ceilDiv(2048, 8) ), 32) ), 2) )}, ithreadIdx.y161{2}, ithreadIdx.x159{32}, iV157{8} ] ca_pos( 3 ) )
T2_l[ iblockIdx.x45{( ceilDiv(( ceilDiv(T0.size[1], 256) ), 2) )}, bblockIdx.y49{( ceilDiv(( ceilDiv(1, 128) ), 2) )}, iS46{2}, bS50{2}, iS51{( ceilDiv(T0.size[0], 32) )}, iS117{( ceilDiv(32, 16) )}, ithreadIdx.z119{( ceilDiv(256, 64) )}, bthreadIdx.y121{( ceilDiv(128, 64) )}, iS123{( ceilDiv(64, 16) )}, bS125{( ceilDiv(64, 16) )}, bS126{16}, ithreadIdx.x190{( ( ( ceilDiv(16, 8) ) * ( ceilDiv(16, 8) ) ) * 8 )}, iS186{8} ] ca_pos( 6 ) produce_pos( 6 )
   = broadcast( T8_l[ iblockIdx.x55{( ceilDiv(( ceilDiv(T0.size[1], 256) ), 2) )}, iS56{2}, iS57{( ceilDiv(T0.size[0], 32) )}, iS127{( ceilDiv(32, 16) )}, ithreadIdx.z129{( ceilDiv(256, 64) )}, iS131{( ceilDiv(64, 16) )}, ithreadIdx.x202{( ( ( ceilDiv(16, 8) ) * ( ceilDiv(16, 8) ) ) * 8 )}, iV200{8} ] ca_pos( 4 ) produce_pos( 3 ) )
T9_l[ iblockIdx.y73{( ceilDiv(( ceilDiv(T1.size[1], 128) ), 2) )}, iS74{2}, iS75{( ceilDiv(T0.size[0], 32) )}, iS133{( ceilDiv(32, 16) )}, ithreadIdx.y135{( ceilDiv(128, 64) )}, iS137{( ceilDiv(64, 16) )}, ithreadIdx.x208{( ( ( ceilDiv(16, 8) ) * ( ceilDiv(16, 8) ) ) * 8 )}, iV204{8} ] ca_pos( 4 ) produce_pos( 3 ) = LdMatrixTranspose( T7_s[ iblockIdx.y79{( ceilDiv(( ceilDiv(T1.size[1], 128) ), 2) )}, iS80{2}, iS81{( ceilDiv(T0.size[0], 32) )}, iS177{( ceilDiv(( ( ( ( ceilDiv(32, 8) ) * ( ceilDiv(8, 1) ) ) * 1 ) * ( ( ( ceilDiv(128, 64) ) * ( ceilDiv(64, 8) ) ) * 8 ) ), 2048) )}, ithreadIdx.z183{( ceilDiv(( ceilDiv(( ceilDiv(2048, 8) ), 32) ), 2) )}, ithreadIdx.y184{2}, ithreadIdx.x182{32}, iV180{8} ] ca_pos( 3 ) )
T3_l[ bblockIdx.x35{( ceilDiv(( ceilDiv(1, 256) ), 2) )}, iblockIdx.y39{( ceilDiv(( ceilDiv(T1.size[1], 128) ), 2) )}, bS36{2}, iS40{2}, iS41{( ceilDiv(T0.size[0], 32) )}, iS107{( ceilDiv(32, 16) )}, bthreadIdx.z109{( ceilDiv(256, 64) )}, ithreadIdx.y111{( ceilDiv(128, 64) )}, bS113{( ceilDiv(64, 16) )}, iS115{( ceilDiv(64, 16) )}, bS114{16}, ithreadIdx.x196{( ( ( ceilDiv(16, 8) ) * ( ceilDiv(16, 8) ) ) * 8 )}, iS192{8} ] ca_pos( 6 ) produce_pos( 6 )
   = broadcast( T9_l[ iblockIdx.y73{( ceilDiv(( ceilDiv(T1.size[1], 128) ), 2) )}, iS74{2}, iS75{( ceilDiv(T0.size[0], 32) )}, iS133{( ceilDiv(32, 16) )}, ithreadIdx.y135{( ceilDiv(128, 64) )}, iS137{( ceilDiv(64, 16) )}, ithreadIdx.x208{( ( ( ceilDiv(16, 8) ) * ( ceilDiv(16, 8) ) ) * 8 )}, iV204{8} ] ca_pos( 4 ) produce_pos( 3 ) )
T5_l[ iblockIdx.x29{( ceilDiv(( ceilDiv(T0.size[1], 256) ), 2) )}, iblockIdx.y31{( ceilDiv(( ceilDiv(T1.size[1], 128) ), 2) )}, iS30{2}, iS32{2}, rS27{( ceilDiv(T0.size[0], 32) )}, rS105{( ceilDiv(32, 16) )}, ithreadIdx.z97{( ceilDiv(256, 64) )}, ithreadIdx.y99{( ceilDiv(128, 64) )}, iS101{( ceilDiv(64, 16) )}, iS103{( ceilDiv(64, 16) )}, iMMA209{( ceilDiv(16, 8) )}, iMMA211{( ceilDiv(16, 8) )}, ithreadIdx.x215{( 8 * ( ceilDiv(8, 2) ) )}, iMMA214{2}, rMMA106{16} ] ca_pos( 4 ) produce_pos( 6 ) = mma(T2_l[ iblockIdx.x45{( ceilDiv(( ceilDiv(T0.size[1], 256) ), 2) )}, bblockIdx.y49{( ceilDiv(( ceilDiv(1, 128) ), 2) )}, iS46{2}, bS50{2}, iS51{( ceilDiv(T0.size[0], 32) )}, iS117{( ceilDiv(32, 16) )}, ithreadIdx.z119{( ceilDiv(256, 64) )}, bthreadIdx.y121{( ceilDiv(128, 64) )}, iS123{( ceilDiv(64, 16) )}, bS125{( ceilDiv(64, 16) )}, bS126{16}, ithreadIdx.x190{( ( ( ceilDiv(16, 8) ) * ( ceilDiv(16, 8) ) ) * 8 )}, iS186{8} ] ca_pos( 6 ) produce_pos( 6 ),T3_l[ bblockIdx.x35{( ceilDiv(( ceilDiv(1, 256) ), 2) )}, iblockIdx.y39{( ceilDiv(( ceilDiv(T1.size[1], 128) ), 2) )}, bS36{2}, iS40{2}, iS41{( ceilDiv(T0.size[0], 32) )}, iS107{( ceilDiv(32, 16) )}, bthreadIdx.z109{( ceilDiv(256, 64) )}, ithreadIdx.y111{( ceilDiv(128, 64) )}, bS113{( ceilDiv(64, 16) )}, iS115{( ceilDiv(64, 16) )}, bS114{16}, ithreadIdx.x196{( ( ( ceilDiv(16, 8) ) * ( ceilDiv(16, 8) ) ) * 8 )}, iS192{8} ] ca_pos( 6 ) produce_pos( 6 ))
}Group End

return false
```

# The solution

I want to remove the above-mentioned check, but we can not just remove that check because this does cause some failures on existing tests. So I am adding a new "fallback mode" where in this new fallback mode, that check is lifted to allow merging. The new "fallback mode" should have lower priority than the existing fallback mode so that the behavior of already-working fusions does not change.

# The refactor

Currently, we have two flags to specify our preference on which group we want to merge: `fallback_mode_enabled_` specifies whether we want to use theorem 4.2 to find mergeable candidates. `preferred_merge_only` specifies whether we want scalar-tensor op to merge. And now we need a new one for whether to lift the consumer checking. However, using boolean flags to specify preference is not very extensible and readable, so I decided to do a refactor of the code.

Now the `sort()` is as simple as
```C++
void ExprSegmentationSorter::sort() {
  Action action = ReducePAAction(nullptr);
  while (!std::holds_alternative<TerminateAction>(action)) {
    action = getNextAction();
    std::visit([this](auto&& act) { do_action(act); }, action);
  }
}
```
where `getNextAction` follow the logic that: Is the most strict version working? If not, then I can relax this constraint. Does this relaxation make it work? If not, I can further relax more constraint. Is it working now? If not, then I can relax more...

This PR is mostly code movement. For already-working fusions, there should be no behavioral change. The only behavioral change is at the places where `relax_consumer_constrain` is used, and this change is only effective to not-already-working fusions like the matmul example above and test `FusionExprSortMatmulLikeSchedule_CUDA`.

### CODE REVIEW

File: third_party/nvfuser/test/test_gpu3.cpp [hunk 1]
```diff
@@ -7352,6 +7352,55 @@ TEST_F(
       "");
 }
 
+TEST_F(NVFuserTest, FusionExprSortMatmulLikeSchedule_CUDA) {
```
Comment (zasdfgbnm) [inline third_party/nvfuser/test/test_gpu3.cpp:7355]: @csarofeen This is the minimum repro for the failing matmul schedule

**Comment by naoyam** [comment]
I'm not sure if it is safe to ignore the constraint even in a fallback case. I think the ID dependency needs to be maintained at least.

Here's what I think we could do:

https://github.com/csarofeen/pytorch/commit/1bd610eb5bd5d1585d45ef0b7cb2a373bd902ac0

All tests including your repro are green.

**Comment by zasdfgbnm** [comment]
> I'm not sure if it is safe to ignore the constraint even in a fallback case. I think the ID dependency needs to be maintained at least.
> 
> Here's what I think we could do:
> 
> [1bd610e](https://github.com/csarofeen/pytorch/commit/1bd610eb5bd5d1585d45ef0b7cb2a373bd902ac0)
> 
> All tests including your repro are green.

I am not sure if I agree or not. My understanding for expr sort is through geometry:  we are happy with merges like:
Screenshot_20230127_161944
But not:
Screenshot_20230127_162114

To do so, why do we have to reject producers like the one below?
Screenshot_20230127_162252
and consumers like the one below?
Screenshot_20230127_162336

They can still have good merges:
Screenshot_20230127_162551

**Comment by csarofeen** [comment]
I think you might be right, but I don't have any kind of proof. I think the latter cases may be fine but what prevents in the latter cases the larger edge from being merged before the smaller?

**Comment by csarofeen** [comment]
Generally the concern has been not about direct consumer producer relationships, but more complex dags where we could merge around one path in a way that prevents merging back around the other.

**Comment by naoyam** [comment]
> Generally the concern has been not about direct consumer producer relationships, but more complex dags where we could merge around one path in a way that prevents merging back around the other.

Yes, `FusionTraversalOrder5` would fail without the constraint as the exprs could be ordered as `tv1->tv2->tv3->tv4-tv5`, but to make the loop generator work, we need `tv1->tv3->tv2->tv4->tv5`.

**Comment by zasdfgbnm** [comment]
@naoyam Could you post your branch as a PR? After some thinking, I feel that your change makes more sense.

**Comment by naoyam** [comment]
@zasdfgbnm See #2385

**Comment by zasdfgbnm** [comment]
Closing in favor of https://github.com/csarofeen/pytorch/pull/2385

**Comment by naoyam** [comment]
@zasdfgbnm Are you going to create a separate cleanup PR? I haven't closely looked at the cleanup part of this PR, but I certainly agree the existing code would need some cleanup.

**Comment by zasdfgbnm** [comment]
> @zasdfgbnm Are you going to create a separate cleanup PR? I haven't closely looked at the cleanup part of this PR, but I certainly agree the existing code would need some cleanup.

Yeah, this makes sense, will do

**Pull request closed without merging**